ksql-cli/src/test/java/io/confluent/ksql/cli/BasicAuthFunctionalTest.java

```diff
@@ -0,0 +1,245 @@
+  private static final String USER_WITHOUT_ACCESS = "tom";
+  private static final String USER_WITHOUT_ACCESS_PWD = "changeme";
+
+  private static final String BASIC_PASSWORDS_FILE_CONTENT =
+      "# Each line generated using org.eclipse.jetty.util.security.Password\n"
+          + USER_WITH_ACCESS + ": changeme,user,developer," + KSQL_CLUSTER_ID + "\n"
+          + USER_WITHOUT_ACCESS + ": MD5:0d107d09f5bbe40cade3de5c71e9e9b7,user,ksql-12\n";
+
+  private static final EmbeddedSingleNodeKafkaCluster CLUSTER = EmbeddedSingleNodeKafkaCluster
+      .newBuilder()
+      .withAdditionalJaasConfig(createJaasConfigContent())
+      .build();
+
+  private static final TestKsqlRestApp REST_APP = TestKsqlRestApp
+      .builder(CLUSTER::bootstrapServers)
+      .withProperty(RestConfig.AUTHENTICATION_METHOD_CONFIG, "BASIC")
+      .withProperty(RestConfig.AUTHENTICATION_REALM_CONFIG, PROPS_JAAS_REALM)
+      .withProperty(RestConfig.AUTHENTICATION_ROLES_CONFIG, KSQL_CLUSTER_ID)
+      .withProperty(JaasUtils.JAVA_LOGIN_CONFIG_PARAM, CLUSTER.getJaasConfigPath())
+      .build();
+
+  @ClassRule
+  public static final RuleChain CHAIN = RuleChain.outerRule(CLUSTER).around(REST_APP);
+
+  @Test
+  public void shouldNotBeAbleToUseWsWithNoCreds() throws Exception {
+    assertThat(makeWsRequest("", ""), is(Code.UNAUTHORIZED));
+  }
+
+  @Test
+  public void shouldBeAbleToUseCliWithValidCreds() throws Exception {
+    assertThat(canMakeCliRequest(USER_WITH_ACCESS, USER_WITH_ACCESS_PWD), is(true));
+  }
+
+  @Test
+  public void shouldBeAbleToUseWsWithValidCreds() throws Exception {
+    assertThat(makeWsRequest(USER_WITH_ACCESS, USER_WITH_ACCESS_PWD), is(HttpStatus.Code.OK));
+  }
+
+  @Test
+  public void shouldNotBeAbleToUseCliWithInvalidPassword() throws Exception {
+    assertThat(canMakeCliRequest(USER_WITH_ACCESS, "wrong pwd"), is(false));
+  }
+
+  @Test
+  public void shouldNotBeAbleToUseWsWithInvalidPassword() throws Exception {
+    assertThat(makeWsRequest(USER_WITH_ACCESS, "wrong pwd"), is(Code.UNAUTHORIZED));
+  }
+
+  @Test
+  public void shouldNotBeAbleToUseCliWithUnknownUser() throws Exception {
+    assertThat(canMakeCliRequest("Unknown-user", "some password"), is(false));
+  }
+
+  @Test
+  public void shouldNotBeAbleToUseWsWithUnknownUser() throws Exception {
+    assertThat(makeWsRequest("Unknown-user", "some password"), is(Code.UNAUTHORIZED));
+  }
+
+  @Test
+  public void shouldNotBeAbleToUseCliWithValidCredsIfUserHasNoAccessToThisCluster() {
+    assertThat(canMakeCliRequest(USER_WITHOUT_ACCESS, USER_WITHOUT_ACCESS_PWD), is(false));
+  }
+
+  @Test
+  public void shouldNotBeAbleToUseWsWithValidCredsIfUserHasNoAccessToThisCluster() throws Exception {
+    assertThat(makeWsRequest(USER_WITHOUT_ACCESS, USER_WITHOUT_ACCESS_PWD), is(Code.UNAUTHORIZED));
+  }
+
+  private boolean canMakeCliRequest(final String username, final String password) {
+    try (KsqlRestClient restClient = new KsqlRestClient(findHttpListener().toString())) {
```

Comment:
final

Reply:
This is `final` by definition. The use of `final` here would be superfluous.